src/Castle.Core.Tests/DynamicProxy.Tests/InParamsTestCase.cs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright 2004-2018 Castle Project - http://www.castleproject.org/
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.DynamicProxy.Tests
+{
+	using System;
+	using System.Linq;
+	using System.Reflection;
+
+	using NUnit.Framework;
+
+	[TestFixture]
+	public class InParamsTestCase : BasePEVerifyTestCase
+	{
+		[Test]
+		public void By_value_parameter_cannot_modify_argument_var()
+		{
+			var x = new ReadOnlyStruct(1);
+			var original = new ReadOnlyStruct(x.Value);
+
+			var different = new ReadOnlyStruct(x.Value + 100);
+			var proxy = generator.CreateInterfaceProxyWithoutTarget<IByValue>(new SetArgumentValueInterceptor(different));
+			proxy.Method(x);
+
+			Assert.AreEqual(original.Value, x.Value);
+		}
+
+		[Test]
+		public void By_referece_In_parameter_cannot_modify_argument_var()
+		{
+			var x = new ReadOnlyStruct(1);
+			var original = new ReadOnlyStruct(x.Value);
+
+			var different = new ReadOnlyStruct(x.Value + 100);
+			var proxy = generator.CreateInterfaceProxyWithoutTarget<IByReadOnlyRef>(new SetArgumentValueInterceptor(different));
+			proxy.Method(in x);
+
+			Assert.AreEqual(original.Value, x.Value);
+		}
+
+		[Test]
+		public void By_reference_Ref_parameter_can_modify_argument_var()
+		{
+			var x = new ReadOnlyStruct(1);
+			var original = new ReadOnlyStruct(x.Value);
+
+			var different = new ReadOnlyStruct(x.Value + 100);
+			var proxy = generator.CreateInterfaceProxyWithoutTarget<IByRef>(new SetArgumentValueInterceptor(different));
+			proxy.Method(ref x);
+
+			Assert.AreNotEqual(original.Value, x.Value);
+		}
+
+		[Test]
+		public void By_reference_Out_parameter_can_modify_argument_var()
+		{
+			var x = new ReadOnlyStruct(1);
+			var original = new ReadOnlyStruct(x.Value);
+
+			var different = new ReadOnlyStruct(x.Value + 100);
+			var proxy = generator.CreateInterfaceProxyWithoutTarget<IOut>(new SetArgumentValueInterceptor(different));
+			proxy.Method(out x);
+
+			Assert.AreNotEqual(original.Value, x.Value);
+		}
+
+		[Test]
+		[TestCase(typeof(IByReadOnlyRef))]
+		public void By_reference_In_parameter_has_ParameterAttributes_In_set(Type type)
```

Comment:
We are mixing facade (or integration) tests here and our expectations about compiler. I'd remove all the tests about expectations, as:
- it's unclear what to do when they start to fail;
- facade tests will start to fail anyway, so these are redundant.

Reply:
Partially agree. I was thinking about whether it would be better to shell out such compiler expectation tests into a separate fixture, but decided against it because it should be obvious that they are there because of, and belong with, the other unit tests. It is also true that if the compiler expectation tests start failing, the actual unit tests will fail, too. In that regard, they are superfluous. However, their added value lies in giving us an additional, very direct hint at *why* the unit tests might be failing. We've done that kind of thing in a few other places recently.

I'll happily remove them but I'd like to leave the final decision to @jonorossi (who was previously in favor of these kinds of tests).

Reply:
I'm definitely in favour of keeping these additional tests that are sort of like proofs, if it happens the implementation of the runtime changes or we misunderstood how it works we'll know straight away. They have heaps of value for DynamicProxy where they wouldn't in most libraries because we are so low in the stack and make so many assumptions about implementation details that a lot of the time aren't documented in the ECMA specifications.

The only thing that would be good is a "heading-like" comment like you've done in the others to separate them from the actual unit tests.

Reply:
Sure, I'll add such a comment!